8254001: [Metrics] Enhance parsing of cgroup interface files for version detection

jerboaa opened the pull request:
This is an enhancement which solves two issues:

1. Multiple reads of relevant cgroup interface files. Now interface files are only read once per file (just like Hotspot).
2. Proxies creation of the impl specific subsystem via `determineType()` as before, but now reads all relevant interface files: `/proc/cgroups`, `/proc/self/mountinfo` and `/proc/self/cgroup`. Once read it passes the parsed information to the impl specific subsystem classes for instantiation. This allows for more flexibility of testing as interface files can be mocked and, thus, more cases can be tested that way without having access to these specific systems. For example, proper regression tests for JDK-8217766 and JDK-8253435 have been added now with this in place.

* [x] Tested on Linux x86_64 on cgroups v1 and cgroups v2. Container tests pass.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8254001](https://bugs.openjdk.java.net/browse/JDK-8254001): [Metrics] Enhance parsing of cgroup interface files for version detection


### Reviewers
 * [Harold Seigel](https://openjdk.java.net/census#hseigel) (@hseigel - **Reviewer**) ⚠️ Review applies to 9a2d6a2095b359a14a801ab6fb76b8cf8f278ac9
 * [Andrew John Hughes](https://openjdk.java.net/census#andrew) (@gnu-andrew - **Reviewer**) ⚠️ Review applies to 078e16db078e32f1595b6e3c1ac3b977efb1d619


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1393/head:pull/1393`
`$ git checkout pull/1393`

jerboaa commented:
@poonamparhar Perhaps you want to take a look at this in the context of not regressing JDK-8255908? Thanks!

jerboaa commented on src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java:
```diff
@@ -146,55 +157,157 @@ static CgroupMetrics create() {
+                        }
+                        case CPUSET_CTRL: {
+                            CgroupInfo info = infos.get(controllerName);
+                            if (info.getMountPoint() != null) {
+                                // On some systems duplicate cpuset controllers get mounted in addition to
+                                // the main cgroup controllers most likely under /sys/fs/cgroup. In that
+                                // case pick the one under /sys/fs/cgroup and discard others.
+                                if (!info.getMountPoint().startsWith("/sys/fs/cgroup")) {
+                                    info.setMountPoint(mountPath);
+                                    info.setMountRoot(mountRoot);
+                                }
+                            } else {
+                                info.setMountPoint(mountPath);
+                                info.setMountRoot(mountRoot);
+                            }
```
This is the actual fix of JDK-8253435 for the Java side.

jerboaa commented on test/jdk/jdk/internal/platform/cgroup/TestCgroupSubsystemFactory.java:
```diff
@@ -190,18 +228,22 @@ public void teardown() {
         assertTrue("Expected non-empty cgroup result", result.isPresent());
         CgroupTypeResult res = result.get();
         assertFalse("Join controller combination expected as cgroups v1", res.isCgroupV2());
+        CgroupInfo memoryInfo = res.getInfos().get("memory");
+        assertEquals("/user.slice/user-1000.slice/session-3.scope", memoryInfo.getCgroupPath());
```
The gist of the Java equivalent of JDK-8253939 (which fixed this for Hotspot). I.e. a regression test for JDK-8217766 and JDK-8254854.

jerboaa commented on test/jdk/jdk/internal/platform/cgroup/TestCgroupSubsystemFactory.java:
```diff
@@ -215,29 +257,39 @@ public void testCgroupv1MultipleCpusetMounts() throws IOException {
         CgroupTypeResult res = result.get();
         assertFalse("Duplicate cpusets should not influence detection heuristic", res.isCgroupV2());
+        CgroupInfo cpuSetInfo = res.getInfos().get("cpuset");
+        assertEquals("/sys/fs/cgroup/cpuset", cpuSetInfo.getMountPoint());
```
We can now assert the proper mount point is being used for multiple cpuset mounts.

poonamparhar commented:
With respect to JDK-8255908, the changes look good to me.

jerboaa commented:
> With respect to JDK-8255908, the changes look good to me.

Thanks!

jerboaa commented:
@bobvandette Please review when you've got some cycles to spare. Much appreciated!

jerboaa commented:
Ping? Anyone?

jerboaa commented:
Anybody willing to review this?

jerboaa commented:
PING?

gnu-andrew commented:
> Anybody willing to review this?

I can have a go.

I have two main concerns:

1. There seems to be little documentation on the new additions. I'm particularly concerned about things like CgroupV1Subsystem.java where a big chunk of documentation on the formats being read is removed and I don't see it being moved elsewhere. Documentation on the new getInstance methods would be helpful too.
2. With the new volatile fields, I think it would be better if the lock being held to initialise them was only held when necessary to perform the assignment. Holding it while performing an extensive process of parsing multiple files that could potentially fail seems a little dangerous. i.e. something like:
if (INSTANCE == null) { CGroupV1Subsystem tmp = initSubSystem(infos); synchronized (CgroupV1Subsystem.class) { if (INSTANCE == null) { INSTANCE = tmp; } } }

jerboaa commented:
> > Anybody willing to review this?
> 
> I can have a go.
> 
> I have two main concerns:
> 
>     1. There seems to be little documentation on the new additions. I'm particularly concerned about things like CgroupV1Subsystem.java where a big chunk of documentation on the formats being read is removed and I don't see it being moved elsewhere. Documentation on the new getInstance methods would be helpful too.

The updated patch now includes some more documentation. The reason why I've removed some of it was because the logic changed and it was outdated.

>     2. With the new volatile fields, I think it would be better if the lock being held to initialise them was only held when necessary to perform the assignment. Holding it while performing an extensive process of parsing multiple files that could potentially fail seems a little dangerous. i.e. something like:
>        if (INSTANCE == null) { CGroupV1Subsystem tmp = initSubSystem(infos); synchronized (CgroupV1Subsystem.class) { if (INSTANCE == null) { INSTANCE = tmp; } } }

I've changed this as suggested, but keep in mind no parsing of multiple files happens after this patch. It happens in CgroupSubsystemFactory.

Thanks,
Severin

jerboaa commented:
@gnu-andrew  Let me know what you think of the updated patch! Thanks.

hseigel reviewed (approved):
Hi Severin,
Thanks for doing this!  Sorry for taking so long to review this change.  The change looks good.  Before pushing it, could you add a comment explaining what the code in lines 185-194 of CgroupSubsystemFactory.java is doing?  Also, please don't overwrite the fix for JDK-8257746.
Thanks again! Harold

jerboaa commented:
/reviewers 2

jerboaa commented:
Setting reviewers to 2 since I want @gnu-andrew to be OK with it too.

jerboaa commented:
Hi Harold,

> Thanks for doing this! Sorry for taking so long to review this change. The change looks good.

Thanks for the review!

> Before pushing it, could you add a comment explaining what the code in lines 185-194 of CgroupSubsystemFactory.java is doing?

Done.

> Also, please don't overwrite the fix for JDK-8257746.

AFAIK this patch doesn't touch this code. So this should not happen (overwrite of the fix). FWIW, once this is in I intend to propose a patch which adds a regression test for JDK-8257746. It would depend on this change.

Thanks,
Severin

gnu-andrew commented:
Thanks, this version looks better.
Sorry for the delay. I didn't realise you'd replied until I received your e-mail. Not sure where notifications for these git PRs end up.

jerboaa commented:
@gnu-andrew Thanks for the review! I'll retest and then integrate.

jerboaa commented:
Tests look good on my end. Also this check passed all Linux tests (it's a no-op everywhere else):
https://github.com/jerboaa/jdk/actions/runs/596859548

jerboaa commented:
/integrate
